pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -60,128 +64,175 @@
+            assert currentSegment.getValidDocIds() != null;
+            currentSegment.getValidDocIds().remove(currentRecordLocation.getDocId());
+            validDocIds.add(recordInfo._docId);
+            return new RecordLocation(immutableSegment, recordInfo._docId, recordInfo._timestamp);
+          } else {
+            return currentRecordLocation;
+          }
         } else {
           // New primary key
           validDocIds.add(recordInfo._docId);
-          return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
+          return new RecordLocation(immutableSegment, recordInfo._docId, recordInfo._timestamp);
         }
       });
     }
     // Update metrics
     _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, _partitionId, ServerGauge.UPSERT_PRIMARY_KEYS_COUNT,
         _primaryKeyToRecordLocationMap.size());
-    return validDocIds;
   }
 
   /**
-   * Updates the upsert metadata for a new consumed record in the given consuming segment.
+   * Updates the upsert metadata for a new consumed record in the given consuming segment. Returns the merged record if
+   * partial-upsert is enabled.
    */
-  public void updateRecord(String segmentName, RecordInfo recordInfo, ThreadSafeMutableRoaringBitmap validDocIds) {
+  public GenericRow updateRecord(IndexSegment segment, RecordInfo recordInfo, GenericRow record) {
+    // For partial-upsert, need to ensure all previous records are loaded before inserting new records.
+    if (_partialUpsertHandler != null) {
+      while (!_partialUpsertHandler.isAllSegmentsLoaded()) {
+        LOGGER
+            .info("Sleeping 1 second waiting for all segments loaded for partial-upsert table: {}", _tableNameWithType);
+        try {
+          //noinspection BusyWait
+          Thread.sleep(1000L);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+
+    _result = record;
     _primaryKeyToRecordLocationMap.compute(recordInfo._primaryKey, (primaryKey, currentRecordLocation) -> {
       if (currentRecordLocation != null) {
         // Existing primary key
 
         // Update the record location when the new timestamp is greater than or equal to the current timestamp. Update
         // the record location when there is a tie to keep the newer record.
         if (recordInfo._timestamp >= currentRecordLocation.getTimestamp()) {
-          currentRecordLocation.getValidDocIds().remove(currentRecordLocation.getDocId());
-          validDocIds.add(recordInfo._docId);
-          return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
+          IndexSegment currentSegment = currentRecordLocation.getSegment();
+          if (_partialUpsertHandler != null) {
+            // Partial upsert
+            GenericRow previousRecord = currentSegment.getRecord(currentRecordLocation.getDocId(), _reuse);
```

Comment:
can `previousRecord ` be null if not found?

Reply:
No, if record location is not null, then we should be able to read the previous record.
`IndexSegment.getRecord()` will never return null

Reply:
hmm what if the segment is expired in between?

Reply:
The entry will be removed before destroying the segment (`removeSegment()` will be called)